[1.17.x] Redstone dust now defers to the blockstate for determining connectivity

Da-Technomancer opened the pull request:
Implements #7915 

Makes redstone dust defer to a new method in IForgeBlockState (which calls a new method in IForgeBlock) to determine whether it should connect to a block on a side. 
The new method in IForgeBlock defaults to the original vanilla logic, so this has no gameplay impact on vanilla features without mods.

Da-Technomancer committed:
Redstone dust now defers to the blockstate for determining connectivity

marchermans reviewed (changes requested):
Please add a test mod showing this feature in action.

marchermans commented on patches/minecraft/net/minecraft/world/level/block/RedStoneWireBlock.java.patch:
```diff
@@ -0,0 +1,11 @@
+          return p_55596_ == p_55595_.m_61143_(ObserverBlock.f_52588_);
+       } else {
+-         return p_55595_.m_60803_() && p_55596_ != null;
++         return p_55595_.canConnectRedstone(p_55596_);
```
A better name would be `canRedstoneConnectTo`

marchermans reviewed (changes requested):
I would choose a different name for the new method.

Da-Technomancer committed:
Changed method name

Changed the name of the new method from canConnectRedstone (the name of the similar method in 1.16) to canRedstoneConnectTo, as suggested.

Da-Technomancer committed:
Added test case

Added a test case, which adds a single block that visually connects to redstone dust specifically on the east side.

Da-Technomancer commented:
I've applied the suggested changes.

Da-Technomancer committed:
Merge branch '1.17.x' into 1.17.x

Da-Technomancer commented:
Did a quick commit to fix the merge conflict on mods.toml. 
This PR should be merge-ready at this point.

rikka0w0 commented:
Is it possible to make this hook be aware of world and blockpos (The old `boolean canConnectRedstone(BlockState state, IBlockReader world, BlockPos pos, @Nullable Direction side)`)? A lot of mod rely on this feature to access the data stored in BlockEntity.

https://github.com/RoyalAliceAcademyOfSciences/SimElectricity/blob/34eb82f68ac2b7c19b45922307cc5e6cac60feae/src/main/java/simelectricity/essential/cable/BlockCable.java#L562

Thanks!

KnightMiner commented:
I would be concerned that making it world sensitive would cause some people to allow redstone to connect but not trigger the proper events to ensure redstone updates. If world/pos access is granted I would add a statement to the javadoc saying the modder is responsible for triggering block updates if the redstone connectivity changes due to something other than the block state properties

That said, I do see little reason not to include redstone connectivity in your blockstate as you probably want to render it in the model as well. The TE can always change the block state as needed, it will persist as long as the block stays the same.

rikka0w0 commented:
> I would be concerned that making it world sensitive would cause some people to allow redstone to connect but not trigger the proper events to ensure redstone updates. If world/pos access is granted I would add a statement to the javadoc saying the modder is responsible for triggering block updates if the redstone connectivity changes due to something other than the block state properties
> 
> That said, I do see little reason not to include redstone connectivity in your blockstate as you probably want to render it in the model as well. The TE can always change the block state as needed, it will persist as long as the block stays the same.

Imagine this scenario, there is a cable block which may have addon panels installed on any sides, there are two catogories of addon panels, one may emit redstone signal, the other does not. Using blockstate, there will be 2^6 combinations for each cable...

Looking at the old `canConnectRedstone`, it provides the world and pos in addition to blockstate and direction. In most case, blockstate should be good enough. But for those really dynamic stuff (like the cable block example I mentioned above), world and blockpos access is still necessary. I would agree that it is the modder's responsiblity to trigger block updates in the latter case.

My proposed solution is presented in #8014.

Da-Technomancer commented:
I agree with the changes by @rikka0w0 in PR #8014, which is based off this PR. 
As a matter of recordkeeping, should I close this PR to avoid having two competing PRs?

ChampionAsh5357 commented:
To address your question, yes it should if you believe the other implementation is better. However, since that has been merged now, I'll close the PR.
